source/MaterialXCore/UnitConverter.cpp

```diff
@@ -0,0 +1,158 @@
+*
+* @file Units.cpp
+* @brief Units support
+*/
+
+
+#include <MaterialXCore/Util.h>
+#include <MaterialXCore/Value.h>
+#include <MaterialXCore/UnitConverter.h>
+
+namespace MaterialX
+{
+
+UnitConverter::UnitConverter(UnitTypeDefPtr unitTypeDef)
+{
+    static const string SCALE_ATTRIBUTE = "scale";
+    static const string OFFSET_ATTRIBUTE = "offset";
+
+    // Populate the unit scale and offset maps for each UnitDef. 
+    vector<UnitDefPtr> unitDefs = unitTypeDef->getUnitDefs();
+    for (UnitDefPtr unitdef : unitDefs)
+    {
+        const string& name = unitdef->getName();
+        if (!name.empty())
+        {
+            const string& scaleString = unitdef->getAttribute(SCALE_ATTRIBUTE);
+            if (!scaleString.empty())
+            {
+                ValuePtr scaleValue = Value::createValueFromStrings(scaleString, getTypeString<float>());
+                _unitScale[name] = scaleValue->asA<float>();
+            }
+            else
+            {
+                _unitScale[name] = 1.0f;
+            }
+
+            const string& offsetString = unitdef->getAttribute(SCALE_ATTRIBUTE);
+            if (!offsetString.empty())
+            {
+                ValuePtr offsetValue = Value::createValueFromStrings(offsetString, getTypeString<float>());
+                _unitOffset[name] = offsetValue->asA<float>();
+            }
+            else
+            {
+                _unitOffset[name] = 1.0f;
+            }
+        }
+    }
+
+    // In case the default unit was not specified in the unittypedef explicit
+    // add this to be able to accept converstion with the default 
+    // as the output unit
+    _defaultUnit = unitTypeDef->getDefault();
+    auto it = _unitScale.find(_defaultUnit);
+    if (it == _unitScale.end())
+    {
+        _unitScale[_defaultUnit] = 1.0f;
+        _unitOffset[_defaultUnit] = 0.0f;
+    }
+}
+
+
+LengthUnitConverter::LengthUnitConverter(UnitTypeDefPtr unitTypeDef) :
+    UnitConverter(unitTypeDef)
+{
+}
+
+UnitConverterPtr LengthUnitConverter::create(UnitTypeDefPtr unitTypeDef)
+{
+    std::shared_ptr<LengthUnitConverter> converter(new LengthUnitConverter(unitTypeDef));
+    return converter;
+}
+
+float LengthUnitConverter::convert(float input, const string& inputUnit, const string& outputUnit) const
+{
+    if (inputUnit == outputUnit)
+    {
+        return 1.0f;
+    }
+
+    auto it = _unitScale.find(inputUnit);
+    if (it == _unitScale.end())
+    {
+        throw ExceptionTypeError("Unrecognized source unit: " + inputUnit);
+    }
+    float fromScale = it->second;
+
+    it = _unitScale.find(outputUnit);
+    if (it == _unitScale.end())
+    {
+        throw ExceptionTypeError("Unrecognized destination unit: " + outputUnit);
+    }
+    float toScale = it->second;
+
+    return (input * fromScale / toScale);
+}
+
+UnitConverterRegistryPtr UnitConverterRegistry::create()
+{
+    std::shared_ptr<UnitConverterRegistry> registry(new UnitConverterRegistry());
```

Comment:
` mx::UnitConverterRegistryPtr registry = mx::UnitConverterRegistry::create();
  registry->addUnitConverter(lengthTypeDef, uconverter);
  uconverter = registry->getUnitConverter(lengthTypeDef);
  REQUIRE(uconverter);
  mx::UnitConverterRegistryPtr registry2 = mx::UnitConverterRegistry::create();
  REQUIRE(registry == registry2);`

`static std::shared_ptr<UnitConverterRegistry> registry(new UnitConverterRegistry());` ?

Reply:
I guess it's the difference between having only 1 registry or allowing any registry to be created. What is the preference?

Reply:
Actually, I think I will just get rid of the registry as it doesn't really serve any purpose. Users can create unit converters as needed.